app/jobs/autoscaling.py

```diff
@@ -198,9 +208,133 @@ def set_desired_capacity(asg, compute_desired_instances, can_scale=True):
     if can_scale:
         if DEBUG:
             print cmd
+        if num_desired < previous_desired:
+            start_draining(asg, previous_desired - num_desired)
+        elif num_desired > previous_desired:
+            stop_draining(asg, num_desired - previous_desired)
         aws_command(cmd)
 
 
+def instances_in(asg):
+    return [item["InstanceId"] for item in asg["Instances"]]
+
+
+def iso_time_now():
+    return datetime.datetime.now().isoformat()
+
+
+def add_draining_tag(instance_ids):
+    cmd = "aws ec2 create-tags --resources {list_instances} --tags Key={draining_tag},Value={timestamp}"
+    cmd = cmd.format(list_instances=' '.join(instance_ids), draining_tag=DRAINING_TAG, timestamp=iso_time_now())
+    aws_command(cmd)
+
+
+def remove_draining_tag(instance_ids):
+    delete_tags_from_instances(instance_ids, [DRAINING_TAG])
+
+
+def start_draining(asg, num_instances):
+    instance_ids = instances_to_drain(asg, num_instances)
+    add_draining_tag(instance_ids)
+
+
+def stop_draining(asg, num_instances):
+    instance_ids = instances_to_rescue(asg, num_instances)
+    add_termination_protection(instance_ids, asg)
```

Comment:
I haven't finished reading this pull request, but it seems odd to be adding a protection here.   At a high level, I would expect protection to be added at creation to all instances, and removed only after they are finished draining (past which point draining cannot be stopped).

Reply:
Yes, that's how it works, see: https://github.com/chanzuckerberg/idseq-infra/pull/148
It is possible to stop draining if the termination has not yet been initiated.
When we want to stop draining, we just need to add protection back and remove the draining tag.
Your comment made me realize I have to add an additional check to instances_to_rescue: filter for state "InService" to guarantee the instance is still usable. Thanks!

Reply:
I think having "protection" implies that thee instance is still usable;  so that place should just make sure protection has not been removed.

Reply:
makes sense, thanks